Enable cross-compilation of SYCL TensorFlow

DuncanMcBain opened the pull request:
These patches add support for crosscompiling TensorFlow with SYCL enabled.
It replaces the cross-compile gcc and g++ with compute++ in sycl-driver
mode, which means that the computecpp.tpl script is no longer needed.
Full instructions on how to use this will be released soon.

DuncanMcBain committed:
First steps to cross compilation

Removed some things that are not needed, started adding definitions
for arm toolchains. computecpp.tpl script is next and needs to be aware
of compiling for the host.

DuncanMcBain committed:
Add correct flags to crosstool

DuncanMcBain committed:
Add temporary compiler hack

DuncanMcBain committed:
Fix broken cxx_flags in crosstool

DuncanMcBain committed:
Fixup Aarch64 toolchain components

DuncanMcBain committed:
Add BUILD.tpl entry for arm cross compiler

DuncanMcBain committed:
Target arm32 with correct directories

DuncanMcBain committed:
Further improvements to the crosscompile toolchain

Now there is a templated bash script that sets the target, gcc-toolchain
and sysroot for compute (the c-like compute++). This means that we no
longer need the version flag, as clang will basically work it out for
us. Similarly, all the builtin directories and include lines have been
removed from the CROSSTOOL as they're not needed any more!

This successfully compiles the host and device objects, but fails to link
because it is not aware that two libComputeCpp.so files are needed - one
for host, and one for device.

DuncanMcBain committed:
Change where ComputeCpp is linked

Now we don't need a host copy of libComputeCpp.so, which makes sense,
as it is only used on the target.

DuncanMcBain committed:
Remove host compiler check in configure script

DuncanMcBain committed:
Fix build config for bazel 0.9.1

DuncanMcBain committed:
Remove compute-script from SYCL config

Using compiler_flag and linker_flag means that C files can see the
target we are using, so the custom compute-script is no longer needed.
This has simplified things considerably!

DuncanMcBain committed:
Add dependency on syclrt to GRU and LSTM ops

Commit:
Removed unneeded code.

Commit:
Set the same build options for both local and cross-build, as much as possible.

Commit:
Added host_crosstool_top to the SYCL lines of tools/bazel.rc, and tidied it up a bit.

Commit:
Renamed ARM_BITCODE_TARGET to BITCODE_TARGET.

Commit:
Fixed build flags from my previous changes.

DuncanMcBain committed:
Merge PR #208 from alistairlow/cross-compilation-cleanup

Flags are unified between the local_linux and crosstool toolchains
Removed unused functions from the configure script
Rename the bitcode option
Add the crosstools to bazel.rc

DuncanMcBain committed:
Fix core kernel tests not building

The kernels didn't have a dependency on the SYCL runtime, which has now
been added (hidden behind the if_sycl function).

DuncanMcBain committed:
Only prompt user for CROSS_TOOLCHAIN if needed

Previously, the sycl_configure file would complain if the env variables
for the cross toolchain couldn't be found - now, it will only ask about
them if the environment variable "CROSS_TOOLCHAIN" is present,
indicating that the user wants to crosscompile TensorFlow + SYCL. The
names have been updated to indicate that the crosscompiling is flexible
enough to work with other GCC-like toolchains, should the need ever
arise.

DuncanMcBain committed:
Fix autodetection of crosscompilation

It now checks the environment in the correct way to see if the user
wishes to crosscompile.

DuncanMcBain committed:
Add Khronos OpenCL loader bazel configuration

ComputeCpp depends on the OpenCL library to communicate with hardware.
Previously, the OpenCL symbols were being resolved dynamically, which
worked well for the Tensorflow wheel, but doesn't for the tests. This
commit adds a minimal bazel configuration for the OpenCL ICD loader
from Khronos and the OpenCL headers, then sets up the dependencies,
such that ComputeCpp depends on the loader, which in turn depends on
the headers. This is more robust as it means that the build depends
less on how the building system is configured.

DuncanMcBain committed:
Improve OpenCL bazel configuration

OpenCL is now a cc_library, meaning that all the kernel tests are linked
against its bits statically. It also uses the exports map from the ICD
repo to indicate versions of the files within it, so should be
compatible with ICD loaders from other vendors. The headers are no
longer an inc_library, as bazel authors indicate that this is not a
useful rule, and library seems to suffice for this case. As of now, the
bazel config will build all tests and the pip package!

jwlawson commented on third_party/sycl/sycl_configure.bzl:
```diff
@@ -175,8 +176,25 @@ def _create_dummy_repository(repository_ctx):
+  if "CROSSTOOL_PYTHON_INCLUDE_PATH" in repository_ctx.os.environ:
+    python_include_path = repository_ctx.os.environ["CROSSTOOL_PYTHON_INCLUDE_PATH"]
+  else:
+    python_include_path = "/usr/include/python2.7"
```
Should this not pull in the headers from the bazel `python_headers` 'library' collection of headers? (as in third_party/toolchains/cpus/py/BUILD)

When crosscompiling you might need to specify other headers, but by default we should probably allow bazel to sort it out for us.

jwlawson commented on third_party/sycl/sycl_configure.bzl:
```diff
@@ -175,8 +176,25 @@ def _create_dummy_repository(repository_ctx):
+    gcc_toolchain_path = ""
+    gcc_toolchain_name = ""
+    opencl_includes = ""
+  spir_type = repository_ctx.os.environ["BITCODE_TARGET"]
```
Where is this specified? Is it at configure time? If not can it be?
I guess this can also apply to all these magical environment variables.

DuncanMcBain replied:
These are all environment variables. I appreciate that's not perfect but for now it's an easy and configurable way to get this stuff working. Ideally, it would be configure-time, or more fixed, but this is early days.

jwlawson commented on third_party/sycl/crosstool/CROSSTOOL.tpl:
```diff
@@ -35,65 +40,175 @@ toolchain {
-  tool_path { name: "strip" path: "/usr/bin/strip" }
+  cxx_flag: "-std=c++11"
+  cxx_flag: "-isystem"
+  cxx_flag: "%{PYTHON_INCLUDE_PATH}%"
```
I don't think that a python include was present before your changes. Is it really needed?

jwlawson commented on tensorflow/core/BUILD:
```diff
@@ -862,6 +862,7 @@ cc_library(
     visibility = ["//visibility:public"],
     deps = [
         ":tensorflow_opensource",
+        "@local_config_sycl//sycl:syclrt",
```
Should this be wrapped in a conditional? if_sycl(...)

DuncanMcBain replied:
Yes, it should be!

DuncanMcBain commented on third_party/sycl/crosstool/CROSSTOOL.tpl:
```diff
@@ -12,6 +12,11 @@ default_toolchain {
   toolchain_identifier: "local_linux"
 }
 
+default_toolchain {
```
Explain why this CPU name is the way it is

jwlawson commented on third_party/sycl/sycl/BUILD.tpl:
```diff
@@ -1,7 +1,7 @@
 
 load("@local_config_sycl//sycl:build_defs.bzl", "if_sycl")
-load("platform", "sycl_library_path")
+load("@local_config_sycl//sycl:platform.bzl", "sycl_library_path")
```
Upstream TF are using 
```
load(":platform.bzl", "sycl_library_path")
load(":platform.bzl", "readlink_command")
```
rather than specifying the whole target in terms of local_config_sycl. Can you uses these instead?

jwlawson reviewed (commented):
I've added a few comments, but my knowledge of bazel is not so great so they may be stupid/make no sense.

By removing the computecpp python script, you so remove the options we have for adding/changing flags at compile time (eg like we use to add the no double flags). This isn't necessarily a problem, but will need us to find another way of adding these flags - probably in the configure script itself to write an action_env in the bazelrc.

I'm not too convinced its a good idea explicitly needing the OpenCL headers and ICD, but if there's no way round it then that's what you have to do.

Overall this looks good though.

lukeiwanski commented on tensorflow/workspace.bzl:
```diff
@@ -390,6 +390,24 @@ def tf_workspace(path_prefix="", tf_repo_name=""):
+  )
+
+  native.new_http_archive(
+      name = "opencl",
```
nitpick: opencl -> opencl-icd?

DuncanMcBain replied:
I can change it to opencl-icd-loader if you want to be totally accurate, sure :)

lukeiwanski commented on third_party/opencl-headers.BUILD:
```diff
@@ -0,0 +1,23 @@
+cc_library(
+    name = "OpenCL-Headers",
+    hdrs = [
+        "opencl22/CL/cl_d3d10.h",
```
are we based of opencl2.2?

DuncanMcBain replied:
The ICD loader builds using the 2.2 sources, so we have to use those

lukeiwanski commented on third_party/sycl/crosstool/CROSSTOOL.tpl:
```diff
@@ -35,65 +40,175 @@ toolchain {
-  # used by gcc. That works because bazel currently doesn't track files at
-  # absolute locations and has no remote execution, yet. However, this will need
-  # to be fixed, maybe with auto-detection?
   cxx_builtin_include_directory: "/usr/lib/gcc/"
```
in principal that can be removed as below line covers it

lukeiwanski reviewed (commented):
It looks all right to me. The only thing that needs to be covered is double / half support ( the thing that causes merge conflict )

jwlawson commented:
Trying this out without wanting to crosscompile I get the following:
```ERROR: no such package '@local_config_sycl//crosstool': key "BITCODE_TARGET" not found in dictionary```
Can you set a default value? This needs to be super easy to use (or at least not harder than the current build commands).

jwlawson commented:
Also seeing
```
ERROR: /tmp/bazel/external/com_google_absl/absl/base/BUILD.bazel:129:1: C++ compilation of rule '@com_google_absl//absl/base:base' failed (Exit 1)
src/main/tools/process-wrapper-legacy.cc:58: "execvp(/usr/local/computecpp/bin/compute, ...)": No such file or directory
```

DuncanMcBain commented:
Instructions have not yet been published, I will get that done so folks can try it out (unfortunately it's not quite so easy as no-configure-go, yet).

jwlawson committed:
[SYCL] Changes SYCL data types configuration (#217)

Set the required flags to disable certain data types for SYCL in the
configure script, directly setting cxxopts in the bazelrc. This means
that these flags are provided without having to set them in the
compiler script.

jwlawson commented on configure.py:
```diff
@@ -1053,13 +1021,12 @@ def main():
   set_action_env_var(environ_cp, 'TF_NEED_OPENCL_SYCL', 'OpenCL SYCL', False)
   if environ_cp.get('TF_NEED_OPENCL_SYCL') == '1':
-    set_host_cxx_compiler(environ_cp)
-    set_host_c_compiler(environ_cp)
```
If these are needed for triSYCL, then these could be moved into the `else` below.

jwlawson commented on configure.py:
```diff
@@ -807,52 +807,6 @@ def set_other_cuda_vars(environ_cp):
-  write_action_env_to_bazelrc('HOST_CXX_COMPILER', host_cxx_compiler)
-
-
-def set_host_c_compiler(environ_cp):
```
Same thing. Is this needed for triSYCL?

jwlawson commented on configure.py:
```diff
@@ -807,52 +807,6 @@ def set_other_cuda_vars(environ_cp):
       write_to_bazelrc('test --config=cuda')
 
 
-def set_host_cxx_compiler(environ_cp):
```
Is this still needed for triSYCL? It looks like the HOST_CXX_COMPILER variable is still being used in https://github.com/lukeiwanski/tensorflow/blob/dev/eigen_mehdi/third_party/sycl/sycl_configure.bzl#L238

jwlawson reviewed (commented):
I just noticed that the triSYCL configure may require HOST_CXX_COMPILER and HOST_C_COMPILER to be set. I don't really know how they are setting things up though, or whether they can just use a much simpler toolchain.
